File: backend/app/models/usage.py
"""
Pydantic models for usage tracking and tier management.
"""

from typing import List, Optional

from pydantic import BaseModel, Field


class UsageStatsResponse(BaseModel):
    """Response model for usage statistics."""

    user_hash: str
    tier: str
    daily_count: int
    daily_limit: int
    daily_remaining: int
    monthly_count: int
    monthly_limit: int
    monthly_remaining: int
    tokens_used_today: int
    tokens_used_month: int
    is_limit_reached: bool
    percentage_used_daily: float
    percentage_used_monthly: float
    reset_daily_at: str
    reset_monthly_at: str


class TierInfoResponse(BaseModel):
    """Response model for tier information."""

    name: str
    daily_limit: int
    monthly_limit: int
    features_enabled: List[str]
    price_monthly: float
    price_yearly: float
    description: str


class AllTiersResponse(BaseModel):
    """Response model for all tier information."""

    tiers: List[TierInfoResponse]
    current_tier: str


class UpgradeTierRequest(BaseModel):
    """Request model for upgrading user tier."""

    tier: str = Field(..., pattern=r"^(free|pro|enterprise)$")


class UsageLimitErrorResponse(BaseModel):
    """Response model for usage limit errors."""

    success: bool = False
    error: str
    tier: str
    limit_type: str  # "daily" or "monthly"
    upgrade_url: str = "/pricing"
    daily_limit: int
    daily_remaining: int
    monthly_limit: int
    monthly_remaining: int

File: types/usage.ts
/**
 * Types for usage tracking and tier management
 */

export type UsageTier = 'free' | 'pro' | 'enterprise'

export interface UsageStats {
  user_hash: string
  tier: UsageTier
  daily_count: number
  daily_limit: number
  daily_remaining: number
  monthly_count: number
  monthly_limit: number
  monthly_remaining: number
  tokens_used_today: number
  tokens_used_month: number
  is_limit_reached: boolean
  percentage_used_daily: number
  percentage_used_monthly: number
  reset_daily_at: string
  reset_monthly_at: string
}

export interface TierInfo {
  name: string
  daily_limit: number
  monthly_limit: number
  features_enabled: string[]
  price_monthly: number
  price_yearly: number
  description: string
}

export interface AllTiersResponse {
  tiers: TierInfo[]
  current_tier: UsageTier
}

export interface UsageCheckResponse {
  success: boolean
  can_generate: boolean
  remaining_today: number
  tier: UsageTier
  daily_limit: number
  monthly_remaining: number
}

export interface UsageLimitError {
  success: false
  error: string
  tier: UsageTier
  limit_type: 'daily' | 'monthly'
  upgrade_url: string
  daily_limit: number
  daily_remaining: number
  monthly_limit: number
  monthly_remaining: number
}

export interface UserFeatures {
  tier: UsageTier
  tier_name: string
  features_enabled: string[]
  bulk_generation_enabled: boolean
  research_enabled: boolean
  api_access_enabled: boolean
}

// Tier display configuration
export const TIER_DISPLAY: Record<UsageTier, {
  name: string
  color: string
  bgColor: string
  borderColor: string
  badgeColor: string
}> = {
  free: {
    name: 'Free',
    color: 'text-gray-700',
    bgColor: 'bg-gray-100',
    borderColor: 'border-gray-200',
    badgeColor: 'bg-gray-500',
  },
  pro: {
    name: 'Pro',
    color: 'text-indigo-700',
    bgColor: 'bg-indigo-100',
    borderColor: 'border-indigo-200',
    badgeColor: 'bg-indigo-500',
  },
  enterprise: {
    name: 'Enterprise',
    color: 'text-purple-700',
    bgColor: 'bg-purple-100',
    borderColor: 'border-purple-200',
    badgeColor: 'bg-purple-500',
  },
}
